### PR TITLE
Enforce use of libraries in thirdparty with option THIRDPARTY=FORCE <master> [9222]

### PR DESCRIPTION
**Pull request by EduPonz**
This PR:

* Adds cmake support for values ON, OFF and FORCE to options `THIRDPARTY` and  `THIRDPARTY_${package}` to force the use of third party packages, regardless of whether they are installed elsewhere.
* Unifies behaviour of `eprosima_find_package` and `eprosima_find_thirdparty`, that is:
    1. Force use of third party packages if `THIRDPARTY=FORCE`
    1. In the case of just `THIRDPARTY=ON`, only use the third party packages if they could not be found elsewhere.
    1. Only update sub-modules if `THIRDPARTY_UPDATE=ON`
* Improves cmake traces.
* Improvement of `FindTinyXML2.cmake`, which now checks for `THIRDPARTY=ON OR THIRDPARTY=FORCE OR THIRDPARTY_TinyXML2=ON/FORCE`, instead of just `THIRDPARTY`.

Please, do not merge until a related PR in https://github.com/eProsima/Fast-RTPS-docs is also ready to merge. I'll only address that when this PR is approved.

### CODE REVIEW

File: cmake/common/eprosima_libraries.cmake [hunk 1]
```diff
@@ -60,40 +102,95 @@ macro(eprosima_find_package package)
...
 
-        if(${package}_FOUND)
-            message(STATUS "${package} library found...")
-        elseif(${FIND_REQUIRED})
-            message(FATAL_ERROR "${package} library not found...")
-        else()
-            message(STATUS "${package} library not found...")
+        # If the package was found we have already show an status message.
+        # if it was not found and it was required throw an fatal error.
+        # if if was not found but it was not required show a warning.
```
Comment (MiguelCompany) [inline cmake/common/eprosima_libraries.cmake:111]:
```suggestion
        # If the package was found we have already shown a status message.
        # if it was not found and it was required, throw a fatal error.
        # if it was not found but it was not required, show a warning.
```

File: cmake/common/eprosima_libraries.cmake [hunk 2]
```diff
@@ -12,45 +12,87 @@
...
 
+# Macro to find cmake built thirdparty libraries.
+#
+# Arguments:
+#   :package: The name of the packge to find. Used for find_package(${package})
+#   :extra arguments: Arguments passed after ${package} are parsed with cmake_parse_arguments()
+#
+# Related CMake options:
+#   :THIRDPARTY: Activate the use of internal thirdparties if external packges is not found [Defaults: OFF]
+#   :THIRDPARTY_FORCE: Force the use of internal thirdparties [Defaults: OFF]. Set to ON if EPROSIMA_BUILD is set to ON.
```
Comment (MiguelCompany) [inline cmake/common/eprosima_libraries.cmake:23]: Looking at the code below, I see that `EPROSIMA_BUILD` sets `THIRDPARTY` to ON, not `THIRDPARTY_FORCE`. Either change the behavior or the documentation.

File: cmake/common/eprosima_libraries.cmake [hunk 3]
```diff
@@ -60,40 +102,95 @@ macro(eprosima_find_package package)
...
 endmacro()
 
+# Macro to find all Fast DDS thirdparty libraries expect for Fast CDR (look at eprosima_find_package).
+#
+# Arguments:
+#   :package: The name of the packge to find. Used for find_package(${package})
+#   :thirdparty_name: The name of the package directory under thirdparty, i.e. thirdparty/${thirdparty_name}
+#
+# Related CMake options:
+#   :THIRDPARTY: Activate the use of internal thirdparties if external packges is not found [Defaults: OFF]
```
Comment (MiguelCompany) [inline cmake/common/eprosima_libraries.cmake:127]:
```suggestion
#   :THIRDPARTY: Activate the use of internal thirdparties if external packges are not found [Defaults: OFF]
```

File: cmake/common/eprosima_libraries.cmake [hunk 4]
```diff
@@ -60,40 +102,95 @@ macro(eprosima_find_package package)
...
+#   :THIRDPARTY_FORCE: Force the use of internal thirdparties [Defaults: OFF]. Set to ON if EPROSIMA_BUILD is set to ON.
```
Comment (MiguelCompany) [inline cmake/common/eprosima_libraries.cmake:128]: Looking at the code below, I see that `EPROSIMA_BUILD` sets `THIRDPARTY` to ON, not `THIRDPARTY_FORCE`. Either change the behavior or the documentation.

File: cmake/common/eprosima_libraries.cmake [hunk 5]
```diff
@@ -12,45 +12,87 @@
...
     if(NOT ${package}_FOUND)
-
         # Parse arguments.
         set(options REQUIRED)
         set(multiValueArgs OPTIONS)
         cmake_parse_arguments(FIND "${options}" "" "${multiValueArgs}" ${ARGN})
 
-        option(THIRDPARTY "Activate the use of internal thirdparties" OFF)
+        # Define options
+        option(THIRDPARTY "Activate the use of internal thirdparties if external packges is not found" OFF)
```
Comment (MiguelCompany) [inline cmake/common/eprosima_libraries.cmake:48]:
```suggestion
        option(THIRDPARTY "Activate the use of internal thirdparties if external packages are not found" OFF)
```

File: cmake/common/eprosima_libraries.cmake [hunk 6]
```diff
@@ -12,45 +12,87 @@
...
         endif()
 
-        option(THIRDPARTY_${package} "Activate the use of internal thirdparty ${package}" OFF)
-
-        if(NOT EPROSIMA_INSTALLER)
+        # 1. If THIRDPARTY_FORCE is set to ON, don't try to find the library outside thirdparty.
+        # 2. For the case of Windows installer, we don't want to look for the package outside thirdparty. This way we
+        #    use thirdparty, meaning we have more control over what is built.
+        if(NOT THIRDPARTY_FORCE AND NOT EPROSIMA_INSTALLER)
+            # Try to quetly find the package outside thridparty first.
```
Comment (MiguelCompany) [inline cmake/common/eprosima_libraries.cmake:62]:
```suggestion
            # Try to quietly find the package outside thridparty first.
```

File: cmake/common/eprosima_libraries.cmake [hunk 7]
```diff
@@ -60,40 +102,95 @@ macro(eprosima_find_package package)
...
+#              at least one of MSVC, MSVC_IDE is set to ON at the same time.
+#   2. If the package is not found in 1), and at least one of THIRDPARTY, THIRDPARTY_FORCE, THIRDPARTY_${package} is set
+#      to ON, use the thirdparty version.
+#         2.1. If THIRDPARTY_UPDATE is set to ON, then update the corresponding git submodule.
+#         2.2. Append the thirdparty source directory to CMAKE_PREFIX_PATH.
+#         2.3. Try to find the package again.
+#   3. If the package was not found anywhere, then print an FATAL_ERROR message.
 macro(eprosima_find_thirdparty package thirdparty_name)
-    if(NOT (EPROSIMA_INSTALLER AND (MSVC OR MSVC_IDE)))
+    option(THIRDPARTY "Activate the use of internal thirdparties if external packges is not found" OFF)
```
Comment (MiguelCompany) [inline cmake/common/eprosima_libraries.cmake:144]:
```suggestion
    option(THIRDPARTY "Activate the use of internal thirdparties if external packages are not found" OFF)
```

File: cmake/common/eprosima_libraries.cmake [hunk 8]
```diff
@@ -60,40 +102,95 @@ macro(eprosima_find_package package)
...
+    if(EPROSIMA_BUILD)
+        set(THIRDPARTY ON)
+    endif()
 
-        option(THIRDPARTY_${package} "Activate the use of internal thirdparty ${package}" OFF)
+    # 1. If THIRDPARTY_FORCE is set to ON, don't try to find the library outside thirdparty.
+    # 2. For the case of Windows installer, we don't want to look for the package outside thirdparty. This way we
+    #    use thirdparty, meaning we have more control over what is built.
+    if(NOT THIRDPARTY_FORCE AND (NOT (EPROSIMA_INSTALLER AND (MSVC OR MSVC_IDE))))
+        # Try to quetly find the package outside thridparty first.
```
Comment (MiguelCompany) [inline cmake/common/eprosima_libraries.cmake:158]:
```suggestion
        # Try to quietly find the package outside thridparty first.
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Apart from my comments below, I think that instead of adding a new option `THIRDPARTY_FORCE` it would be better to have three possible values (`OFF`, `ON` and `FORCE`) for both `THIRDPARTY` and `THIRDPARTY_${package}`

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2551/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2361/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2558/)

**Comment by EduPonz** [comment]
I have taken upon your suggestion of having a three-value option for `THIRDPARTY` and `THIRDPARTY_${package}`. The possible values are `ON`, `OFF` and `FORCE`. `THIRDPARTY_${package}` is set to `THIRDPARTY` unless user specifies otherwise.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2561/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2371/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2568/)

**Review by IkerLuengo — CHANGES_REQUESTED** [review]
Just some mods to make branches easier to read

File: cmake/modules/FindTinyXML2.cmake [hunk 9]
```diff
@@ -20,11 +29,11 @@ if(TinyXML2_FOUND AND NOT THIRDPARTY)
...
-    if(THIRDPARTY OR ANDROID)
+    if(TINYXML2_FROM_THIRDPARTY OR ANDROID)
         set(TINYXML2_FROM_SOURCE ON)
     endif()
 
-    if(THIRDPARTY OR ANDROID)
+    if(TINYXML2_FROM_THIRDPARTY OR ANDROID)
         find_path(TINYXML2_INCLUDE_DIR NAMES tinyxml2.h NO_CMAKE_FIND_ROOT_PATH)
     else()
         find_path(TINYXML2_INCLUDE_DIR NAMES tinyxml2.h)
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply cmake/modules/FindTinyXML2.cmake:39]: :+1:

File: cmake/common/eprosima_libraries.cmake [hunk 10]
```diff
@@ -12,88 +12,232 @@
...
-        if(${package}_FOUND)
-            message(STATUS "${package} library found...")
-        elseif(${FIND_REQUIRED})
-            message(FATAL_ERROR "${package} library not found...")
-        else()
-            message(STATUS "${package} library not found...")
-        endif()
+    # If the package was found we have already show an status message.
+    # if it was not found and it was required throw an fatal error.
+    # if if was not found but it was not required show a warning.
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply cmake/common/eprosima_libraries.cmake:133]: :+1:

File: cmake/common/eprosima_libraries.cmake [hunk 11]
```diff
@@ -12,88 +12,232 @@
...
+    if((NOT ${package}_FOUND) AND ${FIND_REQUIRED})
+        message(FATAL_ERROR "${package} package NOT found")
+    elseif(NOT ${package}_FOUND)
+        message(STATUS "${package} package NOT found")
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply cmake/common/eprosima_libraries.cmake:137]: :+1

**Review by IkerLuengo — APPROVED** [review]
LGTM

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2659/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/70/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2471/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2669/)

**Comment by MiguelCompany** [comment]
Failed test being marked flaky on #1411

**Pull request merged**